server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -0,0 +1,403 @@
+     */
+    public boolean hasValue(Predicate<Object> predicate) {
+        if (leaf == null) {
+            return false;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return false;
+        }
+
+        if (value instanceof List<?> list) {
+            return list.stream().anyMatch(predicate);
+        }
+
+        return predicate.test(value);
+    }
+
+    // Value Update
+
+    /**
+     * Update each value at this path with the {@param transformer} {@link Function}.
+     */
+    @SuppressWarnings("unchecked")
+    public WriteField transform(Function<Object, Object> transformer) {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            ((List<Object>) list).replaceAll(transformer::apply);
+        } else {
+            container.put(leaf, transformer.apply(value));
+        }
+
+        return this;
+    }
+
+    // Value Delete
+
+    /**
+     * Remove all duplicate values from this path.  List order is not preserved.
+     */
+    @SuppressWarnings("unchecked")
+    public WriteField deduplicate() {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            // Assume modifiable list
+            Set<Object> set = new HashSet<>(list);
+            list.clear();
+            ((List<Object>) list).addAll(set);
+        }
+
+        return this;
+    }
+
+    /**
+     * Remove all values at this path that match {@param filter}.  If there is only one value and it matches {@param filter},
+     * the mapping is removed, however empty Lists are retained.
+     */
+    public WriteField removeValuesIf(Predicate<Object> filter) {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            list.removeIf(filter);
+        } else if (filter.test(value)) {
+            container.remove(leaf);
+        }
+
+        return this;
+    }
+
+    /**
+     * Remove the value at {@param index}, if it exists.  If there is only one value and {@param index} is zero, remove the
+     * mapping.
+     */
+    public WriteField removeValue(int index) {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            if (index < list.size()) {
```

Comment:
mutable list assert.

Reply:
As above, an actual `assert` would be turned off and any attempt to modify behavior based on an `UnmodifiableList` works around the scripts intent.  Thoughts?